### PR TITLE
Add initial pass at iterated views (no checking, restricted format)

### PR DESCRIPTION
**Pull request by MattWindsor91**
Since the working copy for this has ended up tangled up in refactors, here is how iterated views looked when Ben and I finished with them.  I'm pushing this to PR now because once I add down closures I imagine the resulting PR would be ridiculously large.

Some things to note:
- There is _no_ down closure checking yet.  I'll add this next.
- Views must be marked as iterated.
- We have some working examples (ARC and single-writer multi-reader) but no idea whether their working-ness is correct or a soundness bug.

### CODE REVIEW

**Comment by MattWindsor91** [comment]
Eep, I left some unit tests dangling.  I'll see if I can fix these next week.

File: Collections.fs [hunk 1]
```diff
@@ -241,6 +241,35 @@ module Multiset =
...
+    /// <typeparam name="State">
+    ///     The type of the accumulator.
+    /// </typeparam>
+    /// <typeparam name="Item">
+    ///     The type of items in the multiset.
+    /// </typeparam>
+    /// <returns>
+    ///     The final value of the state.
+    /// </returns>
+    let fold (f : 'State -> 'Item -> int -> 'State)
```
Comment (septract) [inline Collections.fs:264]: Do you actually use this anywhere (comment from @CaptainHayashi)

File: Flattener.fs [hunk 2]
```diff
@@ -34,26 +33,42 @@ let genFlatFuncSeqName (funcs : Func<'var> seq) : string =
     // This step ensures that the empty view is named 'emp', not 'v'.
     |> fun s -> if s = "v" then "emp" else s
 
+let genFlatIteratedFuncName ifcs =
```
Comment (septract) [inline Flattener.fs:36]: Documentation plz

File: Flattener.fs [hunk 3]
```diff
@@ -34,26 +33,42 @@ let genFlatFuncSeqName (funcs : Func<'var> seq) : string =
...
+    let funcs = Seq.map (fun ifc -> ifc.Func) ifcs
+    genFlatFuncSeqName funcs
+
+let paramsFromIteratedFunc funcContainer =
```
Comment (septract) [inline Flattener.fs:40]: @CaptainHayashi take another look at this - is it used?

File: Flattener.fs [hunk 4]
```diff
@@ -34,26 +33,42 @@ let genFlatFuncSeqName (funcs : Func<'var> seq) : string =
...
+///     A new Func containing all parameters of the individuals as well as their iterators
+///     with the shared variables appended
 /// </returns>
-let flattenFuncSeq
-  (globals : 'var seq)
-  (oview : Func<'var> seq)
-  : Func<'var> =
-    { Name = genFlatFuncSeqName oview
-      Params = oview |> paramsOfFuncSeq |> Seq.append globals |> List.ofSeq }
+let flattenDView : TypedVar seq -> DView -> Func<TypedVar> =
```
Comment (septract) [inline Flattener.fs:59]: Return type should be a DFunc?

File: Flattener.fs [hunk 5]
```diff
@@ -34,26 +33,42 @@ let genFlatFuncSeqName (funcs : Func<'var> seq) : string =
...
+    fun globals dview ->
```
Comment (septract) [inline Flattener.fs:60]: style moan: I'd prefer f# style type annotations, not Haskell-style

File: Flattener.fs [hunk 6]
```diff
@@ -34,26 +33,42 @@ let genFlatFuncSeqName (funcs : Func<'var> seq) : string =
...
+        // TODO: What if iterators share names? e.g. iterated A [n] * iterated B [n]
+        let paramsNoShared = Seq.concat <| Seq.map paramsFromIteratedFunc dview
+        let paramsShared = Seq.toList <| Seq.append paramsNoShared globals
```
Comment (septract) [inline Flattener.fs:63]: weird naming on this line: noshared vs shared.

File: Graph.fs [hunk 7]
```diff
@@ -740,7 +747,7 @@ module Pretty =
         let list = match nk with Normal -> [] | Entry -> [String "(Entry)"] | Exit -> [String "(Exit)"] | EntryExit -> [String "(EntryExit)"]
         hsep [ id |> String
                ([ id |> String
-                  view |> printViewExpr printSVGView ] @ list)
+                  view |> printViewExpr (printIteratedGView (printSym printVar)) ] @ list)
```
Comment (septract) [inline Graph.fs:750]: use GraphView name here?

File: GuardedView.fs [hunk 8]
```diff
@@ -193,6 +218,10 @@ let smgfunc (guard : SMBoolExpr) (name : string) (pars : SMExpr seq) : SMGFunc =
     gfunc guard name pars
 
 
+/// IteratedContainer constructor
+let iterated f it = { Func = f; Iterator = it }
+let emptyIterated f = { Func = f; Iterator = None }
```
Comment (septract) [inline GuardedView.fs:223]: @CaptainHayashi in the wrong file?

Comment (septract) [inline GuardedView.fs:223]: is this used? 
Also should have type annots.

File: GuardedView.fs [hunk 9]
```diff
@@ -319,6 +349,16 @@ let pruneGuardedSet (gset : Multiset<Guarded<_, _>>)
...
 
+/// Gets a Set of TypedVar's from an Expr
+let varsFromExpr : Expr<Sym<Var>> -> Set<TypedVar> =
+    function
+    | Bool e -> mapOverSymVars Mapper.mapBoolCtx findSymVars e
+    | Int e -> mapOverSymVars Mapper.mapIntCtx findSymVars e
+
+/// Gets set of TypedVar's from a SVFunc
+let gfuncVars : SVFunc -> Set<TypedVar> =
+    fun gf -> Set.fold (+) Set.empty (Set.ofList (List.map varsFromExpr gf.Params))
```
Comment (septract) [inline GuardedView.fs:360]: Set.append?

File: GuardedView.fs [hunk 10]
```diff
@@ -327,18 +367,30 @@ let SVGViewVars : GView<Sym<Var>> -> Set<TypedVar> =
...
+            let itemvars = gfuncVars gf
 
-        let gfuncVars gf = Set.fold (+) Set.empty (Set.ofList (List.map symVarExprs gf.Params))
+            condvars
+            + itemvars
 
-        let vars(g, gf) =
-            mapOverSymVars Mapper.mapBoolCtx findSymVars g
-            + gfuncVars gf
+        Set.fold (+) Set.empty (Set.map vars l)
```
Comment (septract) [inline GuardedView.fs:377]: set.append?

File: GuardedView.fs [hunk 11]
```diff
@@ -327,18 +367,30 @@ let SVGViewVars : GView<Sym<Var>> -> Set<TypedVar> =
     fun v ->
```
Comment (septract) [inline GuardedView.fs:367]: Weird structure...

File: GuardedView.fs [hunk 12]
```diff
@@ -327,18 +367,30 @@ let SVGViewVars : GView<Sym<Var>> -> Set<TypedVar> =
...
-        let gfuncVars gf = Set.fold (+) Set.empty (Set.ofList (List.map symVarExprs gf.Params))
+            condvars
+            + itemvars
 
-        let vars(g, gf) =
-            mapOverSymVars Mapper.mapBoolCtx findSymVars g
-            + gfuncVars gf
+        Set.fold (+) Set.empty (Set.map vars l)
+
+let IteratedSVGViewVars : IteratedGView<Sym<Var>> -> Set<TypedVar> =
```
Comment (septract) [inline GuardedView.fs:379]: Could abstract the last two functions?

File: Guarder.fs [hunk 13]
```diff
@@ -17,18 +18,22 @@ open Starling.Lang.Modeller
...
+    let rec guardCFuncIn suffix cview =
+        match cview.Func with
         | CFunc.Func v ->
-            [ { Cond = suffix |> Set.toList |> mkAnd
-                Item = v } ]
+            (* Treat non-iterated views as if they have the iterator [1].
+               This means we don't need to special-case them elsewhere. *)
+            [ { Func = { Cond = suffix |> Set.toList |> mkAnd
+                         Item = v };
+                Iterator = cview.Iterator |> Option.map AVar |> withDefault (AInt 1L) } ]
```
Comment (septract) [inline Guarder.fs:36]: :((

File: Guarder.fs [hunk 14]
```diff
@@ -37,6 +42,7 @@ let guardCView : CView -> GView<Sym<Var>> =
     and guardCViewIn suffix = concatMap (guardCFuncIn suffix)
 
     // TODO(CaptainHayashi): woefully inefficient.
+    let x v =  Multiset.toFlatList v |> guardCViewIn Set.empty
```
Comment (septract) [inline Guarder.fs:45]: rename 'x' plz. Also where does this get used??

File: Main.fs [hunk 15]
```diff
@@ -225,14 +235,20 @@ let printResponse (mview : ModelView) : Response -> Doc =
...
-    | TermGen m -> printVModel (printCmdTerm printSMBoolExpr printSMGView printOView) m
-    | Reify m -> printVModel (printCmdTerm printSMBoolExpr printGuardedSubviewSet printOView) m
+    | TermGen m ->
+        printVModel
+            (printCmdTerm
+                printSMBoolExpr
+                (printIteratedGView (printSym printMarkedVar))
+                printIteratedOView) m
+    | Reify m -> printVModel (printCmdTerm printSMBoolExpr (printSubviewSet printGuardedIteratedSubview) printIteratedOView) m
+    | IterLower m -> printVModel (printCmdTerm printSMBoolExpr (printSubviewSet printGuardedSubview) printOView) m
```
Comment (septract) [inline Main.fs:251]: Ordering in the pipeline?

File: Modeller.fs [hunk 16]
```diff
@@ -674,35 +679,45 @@ let modelDFunc
...
-                                  return Multiset.append lM rM }
+    | ViewSignature.Unit -> ok Multiset.empty
+    | ViewSignature.Func dfunc -> Multiset.map (fun f -> { Func = f; Iterator = None }) <!> (modelDFunc protos dfunc)
+    | ViewSignature.Join(l, r) -> trial { let! lM = modelViewSignature protos l
+                                          let! rM = modelViewSignature protos r
+                                          return Multiset.append lM rM }
+    | ViewSignature.Iterated(dfunc, e) ->
+        let updateFunc (s : string) f = { Func = f; Iterator = Some (Int s) }
+        let modelledDFunc = modelDFunc protos dfunc
+        (Multiset.map (updateFunc e)) <!> modelledDFunc
```
Comment (septract) [inline Modeller.fs:692]: don't know what <!> does

File: Modeller.fs [hunk 17]
```diff
@@ -674,35 +679,45 @@ let modelDFunc
...
+
+let makeIteratorMap : TypedVar option -> VarMap =
+    function
+    | None         -> Map.empty
+    | Some (Int v) -> Map.add v (Type.Int ()) Map.empty
```
Comment (septract) [inline Modeller.fs:697]: map.singleton?

File: Modeller.fs [hunk 18]
```diff
@@ -674,35 +679,45 @@ let modelDFunc
...
+    | _            -> failwith "Iterator in iterated views must be Int type"
 
 /// Produces the environment created by interpreting the viewdef vds using the
 /// view prototype map vpm.
 let rec localEnvOfViewDef vds =
     vds
     |> Seq.ofList
-    |> Seq.map (fun {Params = ps} -> makeVarMap ps)
+    |> Seq.map (fun { Func = {Params = ps}; Iterator = it } -> makeVarMap ps >>= (combineMaps (makeIteratorMap it)))
```
Comment (septract) [inline Modeller.fs:705]: Types plz

File: Modeller.fs [hunk 19]
```diff
@@ -735,21 +750,21 @@ let modelViewDef
 ///         caught them in the main defining view modeller.
 ///     </para>
 /// </remarks>
-let inDefiner : ViewDefiner<SVBoolExpr option> -> Func<'c> list -> bool =
+let inDefiner : ViewDefiner<SVBoolExpr option> -> DView -> bool =
```
Comment (septract) [inline Modeller.fs:753]: lookupInDefiner instead?

File: Modeller.fs [hunk 20]
```diff
@@ -819,7 +834,7 @@ let genAllViewsAt depth (funcs : DFunc seq) : Set<View.Types.DView> =
                 seq { yield []
                       for f in funcs do
                           for e in existing do
-                              yield f :: e }
+                              yield {Iterator = None; Func = f} :: e }
```
Comment (septract) [inline Modeller.fs:837]: Should search generate iterated views?

File: Optimiser.fs [hunk 21]
```diff
@@ -569,18 +569,18 @@ module Graph =
     ///         <c>GView</c>s, but with a <c>BTrue</c> guard.
     ///     </para>
     /// </summary>
-    let (|ITEGuards|_|) (ms: GView<Sym<Var>>)
-      : (BoolExpr<Var> * GView<Sym<Var>>
-         * BoolExpr<Var> * GView<Sym<Var>>) option =
+    let (|ITEGuards|_|) (ms: IteratedGView<Sym<Var>>)
```
Comment (septract) [inline Optimiser.fs:572]: take another look at this? Has been generalised wrt master v.

File: Reifier.fs [hunk 22]
```diff
@@ -12,54 +12,249 @@ open Starling.Core.Model
...
+    // originally by Tomas Petricek: see http://stackoverflow.com/a/16826554
+    seq {
+        match xs with
+        | [] -> yield []
+        | h::t -> for x in powerset t do yield! [x; h::x] }
+
+/// <summary>
+///     Given a series of iterated funcs
+///     <c>A(a)[i] * A(b)[j] * A(c)[k] * ...</c>, generate the equalities
+///     <c>i==j, i==k, ...</c>.
```
Comment (septract) [inline Reifier.fs:41]: a==c, b==c, ...

File: Reifier.fs [hunk 23]
```diff
@@ -12,54 +12,249 @@ open Starling.Core.Model
...
+/// </summary>
+/// <param name="flist">The NON-EMPTY list of funcs to generate over.</param>
+/// <returns>The series of parameter equalities described above.</returns>
+let paramEqualities (flist : IteratedFunc<Sym<MarkedVar>> list)
```
Comment (septract) [inline Reifier.fs:45]: needs tests? These bits are rather subtle.

File: Reifier.fs [hunk 24]
```diff
@@ -12,54 +12,249 @@ open Starling.Core.Model
...
+                else (is, Set.add func.Name ns))
+            (Set.empty, Set.empty)
+
+    (* Now, go through the multiset's funcs.  Since the multiset may contain
+       n copies of a k-iterated func, we normalise to one copy of a k*n-iterated
+       func to get rid of the outer iterator.  What we then do depends on
+       whether the func is iterated or not: if it isn't, we just emit it;
+       otherwise, we add it to an equivalence class based on the func name so
+       we can do the case-split expansion. *)
+    let expandFuncToReify
```
Comment (septract) [inline Reifier.fs:90]: Consider replacing multiset with set now we have iteration in the views? Would make things more regular.

File: Reifier.fs [hunk 25]
```diff
@@ -12,54 +12,249 @@ open Starling.Core.Model
...
+      (ns, ic : Map<string, IteratedGFunc<Sym<MarkedVar>> list>) func n =
+        let norm = TermGen.normalise func n
+        let nname = norm.Func.Item.Name
+        if ifuncs.Contains nname
+        then (ns, Map.add nname (norm::ic.[nname]) ic)
+        else (norm::ns, ic)
+
+    // Have to make sure each class exists first; else exceptions happen.
+    let icempty = ifuncs |> Seq.map (fun name -> (name, [])) |> Map.ofSeq
```
Comment (septract) [inline Reifier.fs:99]: just building an empty instance.

File: Reifier.fs [hunk 26]
```diff
@@ -12,54 +12,249 @@ open Starling.Core.Model
...
+       equalities between the funcs in the class: say, for
+
+       G1->A(a)[i] * G2->A(b)[j] * G3->A(c)[k]
+
+       we have the cases (), (a=b), (a=c), (b=c), (a=b && a=c).  This turns out
+       to be the powerset of the class, less the empty set, where each element
+       of the powerset denotes equality between the members' parameters.
+       We don't need the map keys at this stage. *)
+    let ipsets : (IteratedGFunc<Sym<MarkedVar>> list) seq =
+        iclasses |> Map.toSeq |> Seq.map (snd >> powerset) |> Seq.concat
```
Comment (septract) [inline Reifier.fs:113]: pipe of doom :(

File: Reifier.fs [hunk 27]
```diff
@@ -12,54 +12,249 @@ open Starling.Core.Model
...
+
+    (* Finally, we need to convert those equivalence powersets into a list of
+       funcs. *)
+    let mergeEqualitySet xs =
```
Comment (septract) [inline Reifier.fs:117]: needs a type and a better name for xs (maybe 'state')

Comment (septract) [reply Reifier.fs:117]: add an explicit parameter

File: Reifier.fs [hunk 28]
```diff
@@ -12,54 +12,249 @@ open Starling.Core.Model
...
+                |> List.map iterGFuncTuple
+                |> List.unzip
+            let equalities = paramEqualities funcs
+            let nguard = mkAnd (guards @ equalities)
+
+            // And the new iterator i+j+k+...
+            let iter =
+                funcs
+                |> List.map (fun f -> f.Iterator)
+                |> List.fold mkAdd2 (AInt 0L)
```
Comment (septract) [inline Reifier.fs:136]: should be an explicit sum?

File: Reifier.fs [hunk 29]
```diff
@@ -12,54 +12,249 @@ open Starling.Core.Model
...
+///         any case splits over potentially-equal iterated funcs:
+///         for example, <c>(G1 -> A(x)[i]) * (G2 -> A(y)[j])</c> will also
+///         generate the func <c>(G1 ^ G2 ^ x=y -> A(x)[i+j])</c> when
+///         A(x) is iterated.
+///     </para>
+/// </summary>
+/// <param name="protos">The prototypes used to find iterated funcs.</param>
+/// <param name="view">The view to preprocess.</param>
+/// <returns>The preprocessed view, as a func list.</returns>
+let preprocessView
```
Comment (septract) [inline Reifier.fs:66]: Needs more documentation!

File: Reifier.fs [hunk 30]
```diff
@@ -12,54 +12,249 @@ open Starling.Core.Model
...
+                |> List.map (fun f -> f.Iterator)
+                |> List.fold mkAdd2 (AInt 0L)
+
+            { Iterator = iter
+              Func = { Cond = nguard; Item = (List.head funcs).Func }}
+            :: xs
+
+    Seq.fold mergeEqualitySet nlist ipsets
 
 /// Calculate the multiset of ways that this View matches the pattern in dv and add to the assumulator.
```
Comment (septract) [inline Reifier.fs:144]: 'assumulater'

File: Reifier.fs [hunk 31]
```diff
@@ -12,54 +12,249 @@ open Starling.Core.Model
...
-                  let accumulator =
-                    if p.Name = v.Item.Name && p.Params.Length = v.Item.Params.Length then
-                        matchMultipleViews pattern (rview @ view) accumulator (v::result)
-                    else
-                        accumulator
-                  matchSingleView view (v :: rview) accumulator
-            matchSingleView view [] accumulator
+                | (None, true) ->
+                    (* How many times does a non-iterated func A(x) match an
+                       iterated func (G1->A(y)[n])?  Once, becoming
```
Comment (septract) [inline Reifier.fs:191]: clear this up: should just say 'match a func...'

File: Reifier.fs [hunk 32]
```diff
@@ -12,54 +12,249 @@ open Starling.Core.Model
...
+                    let result =
+                        { Func = func; Iterator = AInt 1L } :: result
+
+                    let view =
+                        { Func = func; Iterator = mkSub2 v.Iterator (AInt 1L) }
+                        :: view
+
+                    matchMultipleViews pattern (rview @ view) accumulator result
+                | (Some n, true) ->
+                    (* How many times does an iterated func A(x)[i]
```
Comment (septract) [inline Reifier.fs:212]: mangled sentence

File: Reifier.fs [hunk 33]
```diff
@@ -12,54 +12,249 @@ open Starling.Core.Model
...
+                    (* We also now match against all of the funcs we
+                       refused earlier (rview). *)
+                    matchMultipleViews pattern (rview @ view) accumulator result
+            (* Now consider the case where we didn't choose the match.
+               This function now goes onto the set of refused funcs that
+               are placed back into any match we do choose. *)
+            matchSingleView p pattern view (v :: rview) accumulator result
+    (* We can now specify what it means to reify a view against an entire
+       view pattern. *)
+    and matchMultipleViews
```
Comment (septract) [inline Reifier.fs:227]: mutually recursive structure is hard to follow!

File: Reifier.fs [hunk 34]
```diff
@@ -12,54 +12,249 @@ open Starling.Core.Model
...
+           cases we can simply traverse the view from left to right, and,
+           every time we find something matching the pattern, split on whether
+           we accept it.
+           Run the case where we do and feed its results into the accumulator,
+           then run the case where we don't with that accumulator. *)
+        | p :: pattern -> matchSingleView p pattern view [] accumulator result
 
     matchMultipleViews dv view accumulator []
 
 /// Reifies an dvs entire view application.
```
Comment (septract) [inline Reifier.fs:243]: @CaptainHayashi says this comment is dumb

File: Collections.fs (hunk 1 above)
Comment (MattWindsor91) [reply Collections.fs:264]: Yes, in Reifier I think.

File: TermGen.fs [hunk 35]
```diff
@@ -18,193 +20,406 @@ open Starling.Core.Command
...
+                   =   (g1 ^ g2 ^ xbar=ybar ^ n>k -> v(xbar)^n-k)       =rsucc
+                     * (g1 ^ !(g2 ^ xbar=ybar)    -> v(xbar)^n)         =rfail
+                     * ((rnext
+                         \ (g2 ^ g1 ^ xbar=ybar ^ k>n -> v(ybar)^k-n))  =qsucc
+                        \  (g2 ^ !(g1 ^ xbar=ybar)    -> v(ybar)^k)))   =qfail
+
+                   Our first task is to decide what each guarded func
+                   (rsucc, rfail, qsucc, qfail) is. *)
+
+                let barEq = List.map2 mkEq xbar ybar |> mkAnd
```
Comment (septract) [inline TermGen.fs:111]: maybe factor out some of the boilerplate below?

File: TermGen.fs [hunk 36]
```diff
@@ -18,193 +20,406 @@ open Starling.Core.Command
...
+/// <param name="r">
+///     The view representing the goal to be subtracted from.
+/// </param>
+/// <param name="q">
+///     The view representing the postcondition to subtract.
+/// </param>
+/// <returns>
+///     The subtracted frame view.
+/// </returns>
 let termGenFrame
```
Comment (septract) [inline TermGen.fs:172]: strange name now?

File: View.fs [hunk 37]
```diff
@@ -40,18 +40,45 @@ module Types =
...
+    type IteratedContainer<'Func, 'Iterator> =
+        { Func : 'Func; Iterator : 'Iterator }
+
+    /// Maps over the item inside an IteratedContainer.
+    let mapIterated (f : 'FuncA -> 'FuncB)
+      ({ Func = v; Iterator = i } : IteratedContainer<'FuncA, 'Var>)
+      : IteratedContainer<'FuncB, 'Var> =
+        { Func = f v; Iterator = i }
+
+    /// Maps over the item inside an IteratedContainer.
```
Comment (septract) [inline View.fs:52]: "map over iterator"

File: starling.fsproj [hunk 38]
```diff
@@ -76,19 +76,25 @@
     <Compile Include="Frontend.fs" />
```
Comment (septract) [inline starling.fsproj:76]: tests need fixing!

File: Flattener.fs (hunk 3 above)
Comment (MattWindsor91) [reply Flattener.fs:40]: Yes, confusingly part of the iterator lowering is happening in `Flattener`.  This would be fairly nontrivial to fix so if ok I'd like to defer until later.

File: GuardedView.fs (hunk 9 above)
Comment (MattWindsor91) [reply GuardedView.fs:360]: Set.unionMany is the correct one I think.  In my changeset-to-come, I got a bit carried away fixing these…

File: Flattener.fs (hunk 4 above)
Comment (MattWindsor91) [reply Flattener.fs:59]: Fix pending

File: Flattener.fs (hunk 5 above)
Comment (MattWindsor91) [reply Flattener.fs:60]: Fix pending (for here and some other areas where I found this happening)

File: Flattener.fs (hunk 6 above)
Comment (MattWindsor91) [reply Flattener.fs:63]: Fix pending: renamed to `allParams`

File: Graph.fs (hunk 7 above)
Comment (MattWindsor91) [reply Graph.fs:750]: Fix pending.

File: GuardedView.fs (hunk 8 above)
Comment (MattWindsor91) [reply GuardedView.fs:223]: Fix pending: also removed `emptyIterated` as nothing uses it.

Comment (MattWindsor91) [reply GuardedView.fs:223]: Nope: see above.  Fix for type annotations pending.

File: GuardedView.fs (hunk 11 above)
Comment (MattWindsor91) [reply GuardedView.fs:367]: I think I've fixed this in my pending changes…?

File: GuardedView.fs (hunk 10 above)
Comment (MattWindsor91) [reply GuardedView.fs:377]: Fix pending

File: GuardedView.fs (hunk 12 above)
Comment (MattWindsor91) [reply GuardedView.fs:379]: I've changed the level of abstraction here so the functions operate over funcs and the callers do the heavy lifting to lift them over views.

File: Guarder.fs (hunk 13 above)
Comment (MattWindsor91) [reply Guarder.fs:36]: Fix pending

File: Guarder.fs (hunk 14 above)
Comment (MattWindsor91) [reply Guarder.fs:45]: Fix pending, and it doesn't: it seems to be a product of an editing slip!

File: Main.fs (hunk 15 above)
Comment (MattWindsor91) [reply Main.fs:251]: This is actually ok, I was misreading things earlier :P

File: Modeller.fs (hunk 16 above)
Comment (MattWindsor91) [reply Modeller.fs:692]: It's `lift`, but I've replaced it with `lift` in my working copy.

File: Modeller.fs (hunk 17 above)
Comment (MattWindsor91) [reply Modeller.fs:697]: Doesn't exist, but I've changed to `Map.ofList`

File: Modeller.fs (hunk 20 above)
Comment (MattWindsor91) [reply Modeller.fs:837]: Unsure, it'll probably be something I look at when I tackle HSF/µZ3

File: Modeller.fs (hunk 19 above)
Comment (MattWindsor91) [reply Modeller.fs:753]: Might leave until later as it's an architecture change

File: Modeller.fs (hunk 18 above)
Comment (MattWindsor91) [reply Modeller.fs:705]: Have now annotated `localEnvOfViewDef` in working copy

File: Optimiser.fs (hunk 21 above)
Comment (MattWindsor91) [reply Optimiser.fs:572]: I _think_ this is ok, but I'm not sure.

File: Reifier.fs (hunk 22 above)
Comment (MattWindsor91) [reply Reifier.fs:41]: Fix pending

File: Reifier.fs (hunk 29 above)
Comment (MattWindsor91) [reply Reifier.fs:66]: Not sure exactly what would be useful here.  Have added a bit more verbiage but I'm not sure if it helps or hinders.

File: Reifier.fs (hunk 24 above)
Comment (MattWindsor91) [reply Reifier.fs:90]: This is probably a thing to do later as it has sweeping architecture changes.

File: Reifier.fs (hunk 25 above)
Comment (MattWindsor91) [reply Reifier.fs:99]: Am making this a bit clearer.

File: Reifier.fs (hunk 26 above)
Comment (MattWindsor91) [reply Reifier.fs:113]: Slated for removal!

File: Reifier.fs (hunk 27 above)
Comment (MattWindsor91) [reply Reifier.fs:117]: Fix pending

File: Reifier.fs (hunk 28 above)
Comment (MattWindsor91) [reply Reifier.fs:136]: I'm adding `mkAdd` for this purpose, though the implementation will be slightly bad for now

File: Reifier.fs (hunk 30 above)
Comment (MattWindsor91) [reply Reifier.fs:144]: Fixing the typo, but I'll leave the comment slightly ugly for now because time constraints.

File: Reifier.fs (hunk 31 above)
Comment (MattWindsor91) [reply Reifier.fs:191]: Fix pending

File: Reifier.fs (hunk 32 above)
Comment (MattWindsor91) [reply Reifier.fs:212]: Fix pending

File: Reifier.fs (hunk 33 above)
Comment (MattWindsor91) [reply Reifier.fs:227]: Not sure what to do here without some significant time sinking :P

File: Reifier.fs (hunk 34 above)
Comment (MattWindsor91) [reply Reifier.fs:243]: Replacing this with a monster of an xml comment :)

File: Reifier.fs (hunk 23 above)
Comment (MattWindsor91) [reply Reifier.fs:45]: I might look at unit tests tomorrow if that's ok?

File: TermGen.fs (hunk 35 above)
Comment (MattWindsor91) [reply TermGen.fs:111]: Fix pending.

File: TermGen.fs (hunk 36 above)
Comment (MattWindsor91) [reply TermGen.fs:172]: Changed it to `termGenWPreMinus` for now, and the function below to `termGenWPre`.  Still unsure.

File: View.fs (hunk 37 above)
Comment (MattWindsor91) [reply View.fs:52]: (Over-)fixed in working copy.

File: starling.fsproj (hunk 38 above)
Comment (MattWindsor91) [reply starling.fsproj:76]: Not sure whether to do this as part of this prq or later?  Will look at tomorrow alongside need for additional tests.
